File: fm-database/fmdb/plan/exercise_screen.py
"""Screen the exercise catalogue against one client record.

The catalogue knows what a movement demands. The client record knows what a
body can take. This module is the join, and it exists because doing that join
by hand is where a coach makes the mistake nobody notices: not choosing a
dangerous exercise, but forgetting that a client with sixteen tagged pain
regions has knees in the list.

FOUR VERDICTS, in descending precedence:

    blocked  — a `block`-severity caution matches the record. Not a ranking
               penalty. The exercise does not get offered.
    caution  — a `caution`-severity caution matches. Offered WITH its
               modification attached; the modification is the point.
    watch    — nothing in the entry matches, but something about this client
               says look twice: a tagged pain region the exercise loads, or a
               balance demand high for their age.
    clear    — nothing fired.

WHY `watch` IS SEPARATE FROM `caution`. A caution is authored knowledge — the
entry says what to do. A watch is derived from the client's own record and has
no authored modification behind it, so presenting the two the same way would
imply guidance that does not exist. The coach decides what a watch means.

Everything here is a NUDGE TO THE COACH, never an automatic edit to a plan.
The one-way door stays hers.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Iterable, Optional

from .contra_screen import _blob  # shared so the negation guard cannot diverge

# ── how the intake body map maps onto the exercise regions ───────────────────
# The intake map is per-side (`knee_left`); exercise entries are side-agnostic
# (`knee`) because an entry cannot know which of a client's knees hurts. Strip
# the side, then fold the map's finer regions onto the coarser exercise set.
_SIDE_SUFFIXES = ("_left", "_right")
_REGION_FOLD = {
    "neck_front": "neck", "neck_back": "neck", "head_back": "neck",
    "scapula": "upper_back",
    "arm": "shoulder",          # the map's "arm" is the upper arm
    "hand": "wrist_hand",
    "upper_abdomen": "abdomen", "lower_abdomen": "abdomen",
    "pelvis": "sacrum_pelvis", "sacrum": "sacrum_pelvis",
    "buttock": "hip",
    "shin": "ankle_foot", "foot": "ankle_foot", "achilles": "ankle_foot",
}
#: Regions on the body map that no exercise entry loads. Folding these onto
#: something would invent an overlap; dropping them is correct.
_REGION_IGNORE = {"head", "face", "jaw"}

#: Client fields that describe what a body can currently take. Passed to
#: `_blob`, which strips negations so "no history of falls" is not read as a
#: falls history.
_SCREEN_FIELDS = (
    "active_conditions", "medical_history", "current_medications", "medications",
    "known_allergies", "notes", "reported_triggers",
    # `weight_loss` is a dict; _blob flattens it to "key value" pairs, which is
    # how `exercise_limitations` and `exercise_current` reach the screen. It is
    # the ONE field explicitly about what a client cannot do, and leaving it out
    # meant a limitation recorded only there ("knee pain (left)") was invisible.
    # The first roster run hid this: the client it would have mattered for also
    # had his condition in active_conditions, so the screen looked right.
    "weight_loss",
)


def fold_pain_regions(pain_locations: Iterable[str]) -> set[str]:
    """Intake body-map slugs → side-agnostic exercise regions."""
    out: set[str] = set()
    for raw in pain_locations or []:
        p = str(raw).strip().lower()
        for suf in _SIDE_SUFFIXES:
            if p.endswith(suf):
                p = p[: -len(suf)]
                break
        if p in _REGION_IGNORE:
            continue
        out.add(_REGION_FOLD.get(p, p))
    return out


@dataclass
class ScreenNote:
    """One reason an exercise is not simply `clear`."""
    kind: str          # "block" | "caution" | "pain" | "age"
    label: str         # the condition or region that fired
    detail: str        # coach-facing explanation
    modification: str = ""


@dataclass
class ExerciseVerdict:
    slug: str
    display_name: str
    client_name: str
    modality: str
    verdict: str                       # blocked | caution | watch | clear
    notes: list[ScreenNote] = field(default_factory=list)
    start_level: Optional[str] = None  # the level label to begin at
    start_reason: str = ""

    @property
    def offerable(self) -> bool:
        return self.verdict != "blocked"


def _age_of(client: dict[str, Any]) -> Optional[int]:
    """Best-effort age. Mirrors Client.estimated_age(): exact if present, else
    the midpoint of the age band."""
    for key in ("estimated_age", "age"):
        v = client.get(key)
        if isinstance(v, int) and v > 0:
            return v
    band = str(client.get("age_band") or "").replace("–", "-").strip()
    if not band:
        return None
    parts = [p for p in band.split("-") if p.strip().isdigit()]
    if len(parts) == 2:
        return (int(parts[0]) + int(parts[1])) // 2
    if len(parts) == 1:
        return int(parts[0])
    return None


def _pick_start_level(exercise: dict[str, Any], supported: bool) -> tuple[Optional[str], str]:
    """Which level to begin at.

    `supported` means something in the screen says keep a hand on something —
    age, a falls signal, an osteoporosis caution. In that case start at the
    first level that names any support, which is how the Otago programme itself
    progresses: supported first, unsupported once recovery strategies are
    confirmed. Otherwise start at the first (easiest) level.
    """
    levels = exercise.get("levels") or []
    if not levels:
        return None, ""
    if supported:
        for lv in levels:
            sup = str(lv.get("support") or "").strip().lower()
            if sup and sup != "none":
                return str(lv.get("level")), "start supported"
    return str(levels[0].get("level")), "start at the easiest level"


def screen_exercise(exercise: dict[str, Any], client: dict[str, Any]) -> ExerciseVerdict:
    """Screen ONE exercise against ONE client record."""
    blob, _asked = _blob(client, _SCREEN_FIELDS)
    pain = fold_pain_regions(client.get("pain_locations") or [])
    age = _age_of(client)

    notes: list[ScreenNote] = []
    blocked = False
    cautioned = False
    supported = False

    for c in exercise.get("cautions") or []:
        terms = [c.get("condition", "")] + list(c.get("condition_aliases") or [])
        if not _any_term_matches(terms, blob):
            continue
        sev = str(c.get("severity") or "caution")
        if sev == "block":
            blocked = True
            notes.append(ScreenNote("block", c.get("condition", ""), c.get("reason", "")))
        else:
            cautioned = True
            supported = True
            notes.append(ScreenNote("caution", c.get("condition", ""),
                                    c.get("reason", ""), c.get("modification", "")))

    # Pain overlap — derived from the client's own body map, no authored
    # modification behind it, so it is a `watch` rather than a caution.
    overlap = sorted(pain & {str(r) for r in (exercise.get("joint_stress") or [])})
    if overlap:
        notes.append(ScreenNote(
            "pain", ", ".join(r.replace("_", " ") for r in overlap),
            "This client has tagged pain in a region the exercise loads. Confirm it is "
            "comfortable before prescribing, and work in a pain-free range.",
        ))

    # Age vs balance demand. The Otago trials found the over-80s benefit MOST
    # from balance work and are also the group a fall costs most, so this is a
    # "start supported" signal, never a reason to withhold.
    demand = int(exercise.get("balance_demand") or 0)
    if age is not None and demand >= 2 and age >= 75:
        supported = True
        notes.append(ScreenNote(
            "age", f"age ~{age}, balance demand {demand}/3",
            "The over-75s gain the most from balance work and can least afford a fall. "
            "Keep support until recovery strategies are confirmed.",
        ))

    verdict = "blocked" if blocked else "caution" if cautioned else "watch" if notes else "clear"
    start_level, start_reason = (None, "") if blocked else _pick_start_level(exercise, supported)

    return ExerciseVerdict(
        slug=exercise.get("slug", "?"),
        display_name=exercise.get("display_name", ""),
        client_name=(exercise.get("client_name") or "").strip() or exercise.get("display_name", ""),
        modality=str(exercise.get("modality") or ""),
        verdict=verdict,
        notes=notes,
        start_level=start_level,
        start_reason=start_reason,
    )


def _any_term_matches(terms: Iterable[str], blob: str) -> bool:
    """Separator-insensitive substring match — mirrors ExerciseCaution.matches().

    Kept here rather than importing the Pydantic method so this module can screen
    plain dicts straight off disk without a model round-trip, which is what the
    shim does.
    """
    import re

    def norm(s: str) -> str:
        return re.sub(r"[^a-z0-9]+", " ", (s or "").lower()).strip()

    hay = norm(blob)
    if not hay:
        return False
    return any(t in hay for t in (norm(x) for x in terms) if t)


_VERDICT_ORDER = {"blocked": 0, "caution": 1, "watch": 2, "clear": 3}


def screen_all(exercises: list[dict[str, Any]], client: dict[str, Any]) -> list[ExerciseVerdict]:
    """Screen the whole catalogue, most-restricted first.

    Blocked entries sort to the top deliberately: the coach should see what is
    off the table before she sees what is on it.
    """
    out = [screen_exercise(e, client) for e in exercises]
    out.sort(key=lambda v: (_VERDICT_ORDER.get(v.verdict, 9), v.modality, v.slug))
    return out


def summarise(verdicts: list[ExerciseVerdict]) -> dict[str, int]:
    counts = {"blocked": 0, "caution": 0, "watch": 0, "clear": 0}
    for v in verdicts:
        counts[v.verdict] = counts.get(v.verdict, 0) + 1
    return counts

File: fm-database-web/src/lib/fmdb/exercise-screen.ts
/**
 * TypeScript mirror of `fmdb/plan/exercise_screen.py`.
 *
 * WHY A MIRROR AND NOT A SHIM. The coach UI can shell out to Python, but the
 * client app cannot — it reads the catalogue off disk. The catalogue already
 * carries three of these pairs (lab-nutrient-priorities, food-cautions,
 * somatic-read) and the house rule is the same each time: the two engines are
 * pinned to a fixture captured from the real Python over the real catalogue,
 * so drift fails a test rather than silently changing what a client is shown.
 *
 * Regenerate the fixture by hand whenever the matcher legitimately changes:
 *   fm-database$ .venv/bin/python -m scripts.dump_exercise_screen_fixture
 */

export type Verdict = "blocked" | "caution" | "watch" | "clear";
export type NoteKind = "block" | "caution" | "pain" | "age";

export interface ScreenNote {
  kind: NoteKind;
  label: string;
  detail: string;
  modification: string;
}

export interface ExerciseVerdict {
  slug: string;
  display_name: string;
  client_name: string;
  modality: string;
  verdict: Verdict;
  notes: ScreenNote[];
  start_level: string | null;
  start_reason: string;
}

/** Mirrors contra_screen._NEGATED. The separator lookahead (not \b) is
 *  load-bearing: intake stores `not_pregnant`, and `_` is a word character. */
const NEGATED =
  /\b(?:not|non|no|never|nil|none|negative|denies|absent|unremarkable)(?=[ _-])(?:[ _-]+(?:a|an|any|the|of|for|h\/o|history|hx|sign|signs|symptom|symptoms|evidence|episode|episodes|known|current|prior|past|significant|reported))*[ _-]+\w+/gi;

/** Mirrors exercise_screen._SCREEN_FIELDS. */
const SCREEN_FIELDS = [
  "active_conditions", "medical_history", "current_medications", "medications",
  "known_allergies", "notes", "reported_triggers",
  // A dict — flattened to "key value" pairs, which is how exercise_limitations
  // and exercise_current reach the screen. See the Python note.
  "weight_loss",
] as const;

const SIDE_SUFFIXES = ["_left", "_right"];
const REGION_FOLD: Record<string, string> = {
  neck_front: "neck", neck_back: "neck", head_back: "neck",
  scapula: "upper_back",
  arm: "shoulder",
  hand: "wrist_hand",
  upper_abdomen: "abdomen", lower_abdomen: "abdomen",
  pelvis: "sacrum_pelvis", sacrum: "sacrum_pelvis",
  buttock: "hip",
  shin: "ankle_foot", foot: "ankle_foot", achilles: "ankle_foot",
};
const REGION_IGNORE = new Set(["head", "face", "jaw"]);

/** Intake body-map slugs → side-agnostic exercise regions. */
export function foldPainRegions(painLocations?: string[]): Set<string> {
  const out = new Set<string>();
  for (const raw of painLocations ?? []) {
    let p = String(raw).trim().toLowerCase();
    for (const suf of SIDE_SUFFIXES) {
      if (p.endsWith(suf)) { p = p.slice(0, -suf.length); break; }
    }
    if (REGION_IGNORE.has(p)) continue;
    out.add(REGION_FOLD[p] ?? p);
  }
  return out;
}

type ClientRecord = Record<string, unknown>;

/** Mirrors contra_screen._blob — flatten the named fields, then strip negations. */
function blob(client: ClientRecord, fields: readonly string[]): string {
  const parts: string[] = [];
  for (const f of fields) {
    if (!(f in client)) continue;
    const v = client[f];
    if (Array.isArray(v)) parts.push(...v.map((x) => String(x)));
    else if (v && typeof v === "object") {
      parts.push(...Object.entries(v as Record<string, unknown>).map(([k, x]) => `${k} ${x}`));
    } else if (v) parts.push(String(v));
  }
  return parts.join(" ").toLowerCase().replace(NEGATED, " ");
}

/** Mirrors ExerciseCaution._normalise — collapse non-alphanumerics to spaces. */
function normalise(s: string): string {
  return (s || "").toLowerCase().replace(/[^a-z0-9]+/g, " ").trim();
}

function anyTermMatches(terms: string[], hay: string): boolean {
  const h = normalise(hay);
  if (!h) return false;
  return terms.map(normalise).filter(Boolean).some((t) => h.includes(t));
}

/** Mirrors exercise_screen._age_of / Client.estimated_age. */
export function estimateAge(client: ClientRecord): number | null {
  for (const key of ["estimated_age", "age"]) {
    const v = client[key];
    if (typeof v === "number" && Number.isInteger(v) && v > 0) return v;
  }
  const band = String(client["age_band"] ?? "").replace("–", "-").trim();
  if (!band) return null;
  const parts = band.split("-").map((p) => p.trim()).filter((p) => /^\d+$/.test(p));
  if (parts.length === 2) return Math.floor((Number(parts[0]) + Number(parts[1])) / 2);
  if (parts.length === 1) return Number(parts[0]);
  return null;
}

interface ExerciseLike {
  slug?: string;
  display_name?: string;
  client_name?: string;
  modality?: string;
  balance_demand?: number;
  joint_stress?: string[];
  levels?: { level?: string; support?: string }[];
  cautions?: {
    condition?: string;
    condition_aliases?: string[];
    severity?: string;
    reason?: string;
    modification?: string;
  }[];
}

function pickStartLevel(ex: ExerciseLike, supported: boolean): [string | null, string] {
  const levels = ex.levels ?? [];
  if (levels.length === 0) return [null, ""];
  if (supported) {
    for (const lv of levels) {
      const sup = String(lv.support ?? "").trim().toLowerCase();
      if (sup && sup !== "none") return [String(lv.level), "start supported"];
    }
  }
  return [String(levels[0].level), "start at the easiest level"];
}

/** Screen ONE exercise against ONE client record. */
export function screenExercise(ex: ExerciseLike, client: ClientRecord): ExerciseVerdict {
  const hay = blob(client, SCREEN_FIELDS);
  const pain = foldPainRegions(client["pain_locations"] as string[] | undefined);
  const age = estimateAge(client);

  const notes: ScreenNote[] = [];
  let blocked = false;
  let cautioned = false;
  let supported = false;

  for (const c of ex.cautions ?? []) {
    const terms = [c.condition ?? "", ...(c.condition_aliases ?? [])];
    if (!anyTermMatches(terms, hay)) continue;
    if ((c.severity ?? "caution") === "block") {
      blocked = true;
      notes.push({ kind: "block", label: c.condition ?? "", detail: c.reason ?? "", modification: "" });
    } else {
      cautioned = true;
      supported = true;
      notes.push({
        kind: "caution", label: c.condition ?? "", detail: c.reason ?? "",
        modification: c.modification ?? "",
      });
    }
  }

  const stress = new Set((ex.joint_stress ?? []).map(String));
  const overlap = [...pain].filter((r) => stress.has(r)).sort();
  if (overlap.length > 0) {
    notes.push({
      kind: "pain",
      label: overlap.map((r) => r.replace(/_/g, " ")).join(", "),
      detail:
        "This client has tagged pain in a region the exercise loads. Confirm it is " +
        "comfortable before prescribing, and work in a pain-free range.",
      modification: "",
    });
  }

  const demand = Number(ex.balance_demand ?? 0);
  if (age !== null && demand >= 2 && age >= 75) {
    supported = true;
    notes.push({
      kind: "age",
      label: `age ~${age}, balance demand ${demand}/3`,
      detail:
        "The over-75s gain the most from balance work and can least afford a fall. " +
        "Keep support until recovery strategies are confirmed.",
      modification: "",
    });
  }

  const verdict: Verdict = blocked
    ? "blocked"
    : cautioned
      ? "caution"
      : notes.length > 0
        ? "watch"
        : "clear";

  const [startLevel, startReason] = blocked ? [null, ""] as [null, string] : pickStartLevel(ex, supported);

  return {
    slug: ex.slug ?? "?",
    display_name: ex.display_name ?? "",
    client_name: (ex.client_name ?? "").trim() || (ex.display_name ?? ""),
    modality: String(ex.modality ?? ""),
    verdict,
    notes,
    start_level: startLevel,
    start_reason: startReason,
  };
}

const VERDICT_ORDER: Record<Verdict, number> = { blocked: 0, caution: 1, watch: 2, clear: 3 };

/** Screen the whole catalogue, most-restricted first. */
export function screenAll(exercises: ExerciseLike[], client: ClientRecord): ExerciseVerdict[] {
  const out = exercises.map((e) => screenExercise(e, client));
  out.sort((a, b) => {
    const d = VERDICT_ORDER[a.verdict] - VERDICT_ORDER[b.verdict];
    if (d !== 0) return d;
    if (a.modality !== b.modality) return a.modality < b.modality ? -1 : 1;
    return a.slug < b.slug ? -1 : a.slug > b.slug ? 1 : 0;
  });
  return out;
}

export function summarise(verdicts: ExerciseVerdict[]): Record<Verdict, number> {
  const counts: Record<Verdict, number> = { blocked: 0, caution: 0, watch: 0, clear: 0 };
  for (const v of verdicts) counts[v.verdict] += 1;
  return counts;
}
